### DISCUSSION TITLE
Better interop with WSL2

### DISCUSSION THREAD

**Discussion by obsgolem** [git-for-windows/git#3715] [Ideas]
WSL2 provides slow access to the Windows filesystem. This is particularly problematic when attempting to use git on largish repos. To combat this, you can install both git for windows and git on linux. That said, the user experience of this solution is not great. To use git for windows from WSL, you by default have to call it via `git.exe`. It would be more convenient if there was a dispatcher which could figure out which filesystem you were on and forward accordingly. Going further, it would be convenient if git for windows could use be set up to automatically use WSL vim, pagers, and rebase tools. Sharing a gitconfig file between versions would be nice as well. I would also like it if I could get a trimmed down version of git for windows which provides windows binaries only for those pieces which would be slow on WSL.

**Comment by dscho** [comment]
> It would be more convenient if there was a dispatcher which could figure out which filesystem you were on and forward accordingly.

Maybe you can write one and share it here?

> it would be convenient if git for windows could use be set up to automatically use WSL vim, pagers, and rebase tools.

Not sure how you want to do that. Maybe you want something like `git config --global core.editor "wsl.exe vim"`?

> Sharing a gitconfig file between versions would be nice as well.

Again, I am not sure how you want to do that. Could you figure it out and share the results here?

> I would also like it if I could get a trimmed down version of git for windows which provides windows binaries only for those pieces which would be slow on WSL.

That would require some work on your part to identify what pieces are slow on WSL.

**Comment by phil-blain** [comment]
> WSL2 provides slow access to the Windows filesystem. 

True, but it does provide fast access to the Linux filesystem. A "solution" could be to clone your repos inside your Linux HOME ?

> Sharing a gitconfig file between versions would be nice as well. 

Since the Windows drive is mounted in WSL, you *could* symlink your WSL `.gitconfig` to your Windows one. It might be hard to keep it 100% compatible between the two OS, but you could put OS-specific bits into separate config files and use [conditional includes](https://git-scm.com/docs/git-config#_conditional_includes) to reference them in your main `.gitconfig`.

**Reply by obsgolem** [reply]
> True, but it does provide fast access to the Linux filesystem. A "solution" could be to clone your repos inside your Linux HOME ?

Unfortunately, many Windows applications don't interact well with the Linux filesystem, for example Visual Studio: https://developercommunity.visualstudio.com/t/MSBuild-project-generated-by-CMake-gives/1567551

**Comment by calle2010** [comment]
I recommend against working across system boundaries with WSL2, especially because of performance, but there are other issues.

But still, if I happen to have my WSL2 terminal open and access a Git repository on the Windows filesystem, I have this in my WSL rc file:

```
function isWinDir {
  case $PWD/ in
    /mnt/*) return $(true);;
    *) return $(false);;
  esac
}
# wrap the git command to either run windows git or linux
function git {
  if isWinDir
  then
    git.exe "$@"
  else
    /usr/bin/git "$@"
  fi
}
```
Depending on your setup, git may run as part of your prompt. I override functions (for ZSH prompt) like __git_prompt_git and parse_git_dirty so that they don't use "command git" or "/usr/bin/git". 

Now I can use WSL git with acceptable performance on Windows filesystem, but it took quite some fiddling to get there.

**Reply by obsgolem** [reply]
Yeah, something like this is what I have ended up doing. It mostly works all right, I am mainly irritated by the need for two copies of config, keys, and rebase tools.

**Comment by mundodisco8** [comment]
I have a similar case, so I would like to add forces to this request.

In my job, we have set up all the dev environment in a docker container, so everybody has the same setup. Code is passed to the container as a volume. In Windows, this means that WSL has to access files that are on the NTFS part of the disk, and compilation times are MADDENING!
I managed to figure out how to keep my code in the EXT part of WSL, and when the repo is passed as a volume to my container, compilation times are back to "native" performance. The problem is that now I can't manage the repo from the Windows side of things. I use Fork as my GUI tool and all hell breaks loose when I try to use it with my WSL-stored repo.

I've just found about [WSLGit](https://github.com/andy-5/wslgit) and so far is doing the trick, but I just started to use it 🤞.

But yeah, a bit more support from Git for Windows would be fantastic! 

Thanks for all the work you do, though!
